nmigen/vendor/quicklogic.py

```diff
@@ -0,0 +1,195 @@
+        "synth",
+        "pack",
+        "place",
+        "route",
+        "write_fasm",
+        "write_bitstream"
+    ]
+    _quicklogic_file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        "{{name}}.v": r"""
+            /* {{autogenerated}} */
+            {{emit_verilog()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.pcf": r"""
+            # {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                set_io {{port_name}} {{pin_name}}
+            {% endfor %}
+        """,
+        "{{name}}.xdc": r"""
+            # {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                {% for attr_name, attr_value in attrs.items() -%}
+                    set_property {{attr_name}} {{attr_value}} [get_ports {{port_name|tcl_escape}} }]
+                {% endfor %}
+            {% endfor %}
+            {{get_override("add_constraints")|default("# (add_constraints placeholder)")}}
+        """,
+        "{{name}}.sdc": r"""
+            # {{autogenerated}}
+            {% for net_signal, port_signal, frequency in platform.iter_clock_constraints() -%}
+                {% if port_signal is not none -%}
+                    create_clock -period {{100000000/frequency}} {{port_signal.name|ascii_escape}}
+                {% endif %}
+            {% endfor %}
+        """
+    }
+    _quicklogic_command_templates = [
+        r"""
+        {{invoke_tool("synth")}}
+            -t {{name}}
+            -v {% for file in platform.iter_extra_files(".v", ".sv", ".vhd", ".vhdl") -%} {{file}} {% endfor %} {{name}}.v
+            -d {{platform.device}}
+            -p {{name}}.pcf
+            -P {{platform.part}}
+            -x {{name}}.xdc
+        """,
+        r"""
+        {{invoke_tool("pack")}}
+            -e {{name}}.eblif
+            -d {{platform.device}}
+            -s {{name}}.sdc
+        """,
+        r"""
+        {{invoke_tool("place")}}
+            -e {{name}}.eblif
+            -d {{platform.device}}
+            -p {{name}}.pcf
+            -n {{name}}.net
+            -P {{platform.part}}
+            -s {{name}}.sdc
+        """,
+        r"""
+        {{invoke_tool("route")}}
+            -e {{name}}.eblif
+            -d {{platform.device}}
+            -s {{name}}.sdc
+        """,
+        r"""
+        {{invoke_tool("write_fasm")}}
+            -e {{name}}.eblif
+            -d {{platform.device}}
+            -s {{name}}.sdc
+        """,
+        r"""
+        {{invoke_tool("write_bitstream")}}
+            -f {{name}}.fasm
+            -d {{platform.device}}
+            -P {{platform.part}}
+            -b {{name}}.bit
+        """
+    ]
+
+    # Common logic
+
+    def __init__(self, *):
+        super().__init__()
+
+    @property
```

Comment:
None of the following four functions are necessary--just make them normal properties. They only need to be functions with multiple supported toolchains.